src/compiler/checker.ts

```diff
@@ -34515,31 +34515,35 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkAssertionWorker(node: JSDocTypeAssertion | AssertionExpression, checkMode: CheckMode | undefined) {
-        let type: TypeNode;
+        let typeNode: TypeNode;
         let expression: Expression;
         switch (node.kind) {
             case SyntaxKind.AsExpression:
             case SyntaxKind.TypeAssertionExpression:
-                type = node.type;
+                typeNode = node.type;
                 expression = node.expression;
                 break;
             case SyntaxKind.ParenthesizedExpression:
-                type = getJSDocTypeAssertionType(node);
+                typeNode = getJSDocTypeAssertionType(node);
                 expression = node.expression;
                 break;
         }
         const exprType = checkExpression(expression, checkMode);
-        if (isConstTypeReference(type)) {
+        if (isConstTypeReference(typeNode)) {
             if (!isValidConstAssertionArgument(expression)) {
                 error(expression, Diagnostics.A_const_assertions_can_only_be_applied_to_references_to_enum_members_or_string_number_boolean_array_or_object_literals);
             }
             return getRegularTypeOfLiteralType(exprType);
         }
-        const links = getNodeLinks(node);
-        links.assertionExpressionType = exprType;
-        checkSourceElement(type);
-        checkNodeDeferred(node);
-        return getTypeFromTypeNode(type);
+        checkSourceElement(typeNode);
+        const type = getTypeFromTypeNode(typeNode);
+        // See if we need to check whether the expression and asserted types are comparable.
+        if (!isErrorType(type) && !((exprType.flags | type.flags) & (TypeFlags.AnyOrUnknown | TypeFlags.Never))) {

```

Comment:
IMO, this is silly - this is quite literally what the first lines of `isTypeAssignableTo` does, and all doing it early avoids is putting a node id into a set (in exchange for duplicating this comparison in the cases where that _isn't_ avoided, which is far and away more common).